plumpy/events.py

```diff
@@ -3,13 +3,42 @@
 import sys
 import asyncio
 
-__all__ = ['new_event_loop', 'set_event_loop', 'get_event_loop', 'run_until_complete']
+__all__ = [
+    'new_event_loop', 'set_event_loop', 'get_event_loop', 'run_until_complete', 'set_event_loop_policy',
+    'reset_event_loop_policy', 'PlumpyEventLoopPolicy'
+]
 
 get_event_loop = asyncio.get_event_loop  # pylint: disable=invalid-name
 new_event_loop = asyncio.new_event_loop  # pylint: disable=invalid-name
 set_event_loop = asyncio.set_event_loop  # pylint: disable=invalid-name
 
 
+class PlumpyEventLoopPolicy(asyncio.DefaultEventLoopPolicy):
+    """Custom event policy that always returns the same event loop that is made reentrant by ``nest_asyncio``."""
+
+    _loop = None
+
+    def get_event_loop(self):
```

Comment:
I'm not sure if we'll overwrite `new_event_loop` to do the same thing as this.  I'm thinking about the case of a child runner where, I guess, it calls `new_event_loop` when the runner gets instantiated.  We could, of course, change the runner code but it feels like whether the loop is reentrant or not should be transparent to clients (at least as much as possible).  One problem with this is that if a client calls `new_event_loop` there is no way for us to know if they were just planning to run a new loop in serial after destroying the old one, or to, as in AiiDA, they were planning to run the loop inside the current one...

Reply:
I am not sure either. For what it's worth, neither in `plumpy`'s nor in `aiida-core`'s source code, is `new_event_loop` ever called. That doesn't help with the problem for other users of this library of course. Though I am not sure if `plumpy` is even being used outside of `aiida-core`.

Reply:
@muhrin Coming back to this, because ideally we need to finish this up a.s.a.p. If the current loop is destroyed before they call `new_event_loop` they will get a new reentrant loop, even if we override it to behave exactly as our `get_event_loop`, wouldn't it? Because if they delete it, the global loop that `get_event_loop` would return also no longer exists and so a new one is created. The problem, however, might be what happens when they call `new_event_loop` without having deleted the current one, if we override `new_event_loop` to be the same as `get_event_loop`, then that behavior might be surprising. Maybe we should just document that `plumpy` works this way though and only ever has a single reentrant loop. Maybe even explicitly raise `NotImplementedError` for `new_event_loop`? Anyway, since this is also not relevant for `aiida-core` for now, and `plumpy` does not really have other users, maybe we should make a choice and release it. If we later on find out it is not ideal for a concrete reason we can fix it.

Reply:
> Maybe even explicitly raise `NotImplementedError` for `new_event_loop`?

I vote for this. We do not need to worry about the `new_event_loop` behaviour if it not be used anywhere else at the moment, isn't it? And raise `NotImplementedError` to avoid strange behaviour from happening in the future.

Reply:
Sorry for the delay - there have been a lot of meeting recently.

OK, so my suggestion about what `new_event_loop` should do was based on a misunderstanding on my part on what `asyncio` does natively. I've just rechecked the code and actually, `new_event_loop` should probably just return a new instance of a reentrant event loop.  But I'm also fine with the `NotImplementedError` given that it's not used.

Reply:
If the `new_event_loop` normally does simply destroy the global loop and then create a new one, we could simply replicate this, couldn't we? What happens with existing references to the original loop though? Is this is also what happens with vanilla `asyncio`? Will `new_event_loop` actually destroy some existing event loop instances? I thought the idea was there that you could have multiple event loops but just not have them nested?

Reply:
Well, [normally](https://github.com/python/cpython/blob/master/Lib/asyncio/events.py#L653), `new_event_loop` does nothing more than instantiate a new loop.  Doesn't destroy anything or set that new loop to be the current loop so if we do the same then we're fulfilling the contract.  And, yes, you're right you can have multiple, non-nested, loops (which should also work with multiple instance of reentrant loops).